[JSC] Use fast iteration for array destructuring

Constellation opened the pull request:
#### e12d80c03d08d0cb49b0d952823c002c832e5ecf
<pre>
[JSC] Use fast iteration for array destructuring
<a href="https://bugs.webkit.org/show_bug.cgi?id=246672">https://bugs.webkit.org/show_bug.cgi?id=246672</a>
rdar://101275126

Reviewed by NOBODY (OOPS!).

This patch integrates fast for-of iteration into Array destructuring so that we do not allocate iteration result objects for
optimized array iteration case. We already integrated it into for-of, but we should do that for array destructuring too.
This patch improves the attached microbenchmark by 2x.

                                                  ToT                     Patched

map-iteration-and-array-destructuring       24.9895+-0.1555     ^     12.4263+-0.0735        ^ definitely 2.0110x faster

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitEnumeration):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ArrayPatternNode::bindValue const):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e12d80c03d08d0cb49b0d952823c002c832e5ecf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93255 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2452 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23901 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102945 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163230 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2459 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30767 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85633 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99060 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98925 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1709 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79717 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28617 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83622 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83381 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71730 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/is-web-process-responsive (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84572 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37156 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17259 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79637 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34977 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18502 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27529 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38850 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41036 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82270 "Built successfully") | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1827 "Failed to update pull request") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40780 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37716 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18589 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

shvaikalesh reviewed (approved):
This is a perfection, r=me. Could you please run `test/language/expressions/assignment/dstr` of test262 suite to ensure there is no regression? We have quite a few failures there already but IIRC those are related to `let [x, y] = [1, 0]` shortcut.

Constellation committed:
[JSC] Use fast iteration for array destructuring
https://bugs.webkit.org/show_bug.cgi?id=246672
rdar://101275126

Reviewed by Alexey Shvayka.

This patch integrates fast for-of iteration into Array destructuring so that we do not allocate iteration result objects for
optimized array iteration case. We already integrated it into for-of, but we should do that for array destructuring too.
This patch improves the attached microbenchmark by 2x.

                                                  ToT                     Patched

map-iteration-and-array-destructuring       24.9895+-0.1555     ^     12.4263+-0.0735        ^ definitely 2.0110x faster

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitEnumeration):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ArrayPatternNode::bindValue const):

Canonical link: https://commits.webkit.org/255681@main
